### PR TITLE
[App Provider] Add features prop

### PR DESCRIPTION
**Pull request by tmlayton**
### WHY are these changes introduced?

We want a way to toggle certain features within Polaris React. 

### WHAT is this pull request doing?

Adding a `features` prop to app provider and creating a hook for settings context.

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [x] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

### CODE REVIEW

File: src/components/AppProvider/AppProvider.tsx [hunk 1]
```diff
@@ -95,25 +98,27 @@ export class AppProvider extends React.Component<AppProviderProps, State> {
   }
 
   render() {
-    const {theme = {logo: null}, children} = this.props;
+    const {theme = {logo: null}, settings, children} = this.props;
```
Comment (danrosenthal) [inline src/components/AppProvider/AppProvider.tsx]: Do we know if children will receive an update when this changes?

Comment (tmlayton) [reply src/components/AppProvider/AppProvider.tsx]: Yep, looks like it.

File: src/components/AppProvider/AppProvider.tsx [hunk 2]
```diff
@@ -36,6 +37,8 @@ export interface AppProviderProps extends AppBridgeOptions {
   linkComponent?: LinkLikeComponent;
   /** Custom logos and colors provided to select components */
   theme?: Theme;
+  /** For toggling settings */
+  settings?: Settings;
```
Comment (tmlayton) [inline src/components/AppProvider/AppProvider.tsx]: Are we good with this prop name? Other options could be `features`, `flags`, etc.

Comment (danrosenthal) [reply src/components/AppProvider/AppProvider.tsx]: `flags` might be a better indicator of its intended use, but I'm comfortable with `settings`

Comment (BPScott) [reply src/components/AppProvider/AppProvider.tsx]: bikeshedding: flags or features feel better to me - "betas" while being our internal name for some of this stuff implies unfinished while settings feels a bit too permanent.

File: src/utilities/settings/hooks.tsx [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import {useContext} from 'react';
+import {SettingsContext} from './context';
+
+export function useSettings() {
+  return useContext(SettingsContext);
```
Comment (BPScott) [inline src/utilities/settings/hooks.tsx]: Throw an exception if the context is falsy here - see the other hooks like sticky-manager for an example.

That'll ensure that this hook returns a value, and you won't need to do `settings != null` checks whenever you use the result of this hook

File: src/utilities/settings/types.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export interface Settings {
+  [key: string]: boolean;
```
Comment (BPScott) [inline src/utilities/settings/types.ts]: Can this have an explicit shape? Telling consumers "you can toss anything into this object" doesn't feel that helpful when it comes to showing what's available.

Comment (danrosenthal) [reply src/utilities/settings/types.ts]: How much more explicit could you get without limiting flexibility? Or would you see us opening up the type each time we add a new key?

Comment (tmlayton) [reply src/utilities/settings/types.ts]: Chatted directly about this

Comment (BPScott) [reply src/utilities/settings/types.ts]: Context on that chat: I was thinking adding to the type each time you add a new key - doing that feels like a cheap price to pay to ensure we get "isAdafafaf is not a valid key" warnings if we do stuff like `{isAdafafaf} = useSettings();`.

But Tim made a point that we want to be obtuse about this for REASONS for the moment so I'm happy with this.

**Pull request merged**